<a href="https://colab.research.google.com/github/sarahcodebyte/NIRF-Rank-Predictor--Hack-A-Duck/blob/main/NIRF_Rank_Predictor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

This project is made to judge data analysis and machine learning skills at beginner level. Two datasets are used to do the same. 

This project is created by the contributions of Sarah Shruti Kujur and Arpan Samuel Topno for the competition Hack-A-Duck 2022, organized by Society of Data Science, BIT Mesra.

# **Importing Dependencies**

In [1]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import scale
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix


In [2]:
pd.options.plotting.backend = "plotly"

# **Data Collection and Data Preprocessing**

In [3]:
dataset1 = pd.read_csv('/content/EngineeringRanking.csv')

In [4]:
dataset1.head()

Institute Id                            Institute Name       City  \
0  IR-E-U-0456     Indian Institute of Technology Madras    Chennai   
1  IR-E-I-1074      Indian Institute of Technology Delhi  New Delhi   
2  IR-E-U-0306     Indian Institute of Technology Bombay     Mumbai   
3  IR-E-I-1075     Indian Institute of Technology Kanpur     Kanpur   
4  IR-E-U-0573  Indian Institute of Technology Kharagpur  Kharagpur   

           State  Score_21  Rank_21  TLR_21  RPC_21  GO_21  OI_21  ...  GO_17  \
0     Tamil Nadu     90.19      1.0   95.47   96.43  81.92  62.44  ...  83.78   
1          Delhi     88.96      2.0   91.76   95.82  80.97  66.39  ...  77.45   
2    Maharashtra     85.16      3.0   89.32   92.56  79.71  53.68  ...  83.64   
3  Uttar Pradesh     83.22      4.0   86.71   83.13  89.79  55.96  ...  74.29   
4    West Bengal     82.03      5.0   80.51   88.59  83.01  58.46  ...  88.02   

   OI_17  Perception_17  Score_16  Rank_16  TLR_16  RPC_16  GO_16  OI_16  \
0  77.19          81.46     89.41        1   88.26   94.02  81.81  86.11   
1  71.41          77.24     82.02        4   80.27   91.62  74.72  66.17   
2  69.70          84.24     87.66        2   85.93   94.14  84.97  74.84   
3  61.35          73.59     81.07        5   66.08   93.52  85.62  70.59   
4  74.11          73.43     83.91        3   76.23   92.68  83.95  78.05   

   Perception_16  
0           98.0  
1           98.0  
2           99.0  
3           98.0  
4           97.0  

[5 rows x 46 columns]

In [5]:
dataset1.isnull().value_counts() #checking if there are any null values

Institute Id  Institute Name  City   State  Score_21  Rank_21  TLR_21  RPC_21  GO_21  OI_21  Perception_21  Score_20  Rank_20  TLR_20  RPC_20  GO_20  OI_20  Perception_20  Score_19  Rank_19  TLR_19  RPC_19  GO_19  OI_19  Perception_19  Score_18  Rank_18  TLR_18  RPC_18  GO_18  OI_18  Perception_18  Score_17  Rank_17  TLR_17  RPC_17  GO_17  OI_17  Perception_17  Score_16  Rank_16  TLR_16  RPC_16  GO_16  OI_16  Perception_16
False         False           False  False  False     False    False   False   False  False  False          False     False    False   False   False  False  False          False     False    False   False   False  False  False          False     False    False   False   False  False  False          False     False    False   False   False  False  False          False     False    False   False   False  False  False            48
                                                                                                                                            

In [6]:
dataset1.nunique()

Institute Id      273
Institute Name    273
City              142
State              30
Score_21          191
Rank_21           191
TLR_21            190
RPC_21            193
GO_21             193
OI_21             196
Perception_21      72
Score_20          195
Rank_20           195
TLR_20            193
RPC_20            197
GO_20             196
OI_20             195
Perception_20      79
Score_19          185
Rank_19           185
TLR_19            192
RPC_19            193
GO_19             197
OI_19             191
Perception_19      81
Score_18           98
Rank_18            98
TLR_18             99
RPC_18             99
GO_18             100
OI_18              99
Perception_18      65
Score_17           97
Rank_17            97
TLR_17            100
RPC_17             99
GO_17              99
OI_17              98
Perception_17     100
Score_16           81
Rank_16            81
TLR_16             81
RPC_16             82
GO_16              82
OI_16              81
Perception

In [7]:
dataset2 = pd.read_csv('/content/nirf_2022_EngineeringRanking_ranking.csv')

In [8]:
dataset2.head()

Institute ID                                       Name       City  \
0  IR-E-U-0456      Indian Institute of Technology Madras    Chennai   
1  IR-E-I-1074      Indian Institute of Technology, Delhi  New Delhi   
2  IR-E-U-0306     Indian Institute of Technology, Bombay     Mumbai   
3  IR-E-I-1075      Indian Institute of Technology Kanpur     Kanpur   
4  IR-E-U-0573  Indian Institute of Technology, Kharagpur  Kharagpur   

           State  Score  Rank  
0     Tamil Nadu  90.04     1  
1          Delhi  88.12     2  
2    Maharashtra  83.96     3  
3  Uttar Pradesh  82.56     4  
4    West Bengal  78.89     5

In [9]:
dataset2.isnull().value_counts() #checking if there are any null values

Institute ID  Name   City   State  Score  Rank 
False         False  False  False  False  False    200
dtype: int64

In [10]:
dataset2.nunique()

Institute ID    200
Name            199
City            118
State            30
Score           191
Rank            191
dtype: int64

In [11]:
ml_dataset = dataset2[['Institute ID', 'Score', 'Rank']] #dataset used for creating a ML model

# **Data Analysis and Visualization**

Data Analysis done on Dataset1

In [12]:
df_scores = dataset1[['Institute Name', 'Score_21', 'Score_20', 'Score_19', 'Score_18', 'Score_17', 'Score_16']]

In [13]:
df_scores.loc[:,'Nan'] = df_scores.isnull().sum(axis=1)

df_scores = df_scores[df_scores['Nan'] < 4]

df_scores.reset_index(inplace=True, drop=True)

df_scores = df_scores.drop(columns='Nan', axis=1)

/usr/local/lib/python3.8/dist-packages/pandas/core/indexing.py:1667: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[key] = value


In [14]:
df_scores = df_scores[['Institute Name', 'Score_16', 'Score_17', 'Score_18', 'Score_19', 'Score_20', 'Score_21']]

df_scores.head()

Institute Name  Score_16  Score_17  Score_18  \
0     Indian Institute of Technology Madras     89.41     87.96     88.95   
1      Indian Institute of Technology Delhi     82.02     81.08     82.18   
2     Indian Institute of Technology Bombay     87.66     87.87     84.82   
3     Indian Institute of Technology Kanpur     81.07     76.83     75.24   
4  Indian Institute of Technology Kharagpur     83.91     81.93     77.78   

   Score_19  Score_20  Score_21  
0     89.05     89.93     90.19  
1     85.36     88.08     88.96  
2     84.40     85.08     85.16  
3     77.57     82.18     83.22  
4     79.41     80.56     82.03

***Top 10 Institutes***

In [15]:
fig = df_scores.head(10).set_index('Institute Name').T.plot()
fig.update_layout(width=1000, height=500)
fig.show()

***Bottom 10 Institutes***

In [16]:
fig = df_scores.tail(10).set_index('Institute Name').T.plot()
fig.update_layout(width=1000, height=500)
fig.show()

***Performance of IITs***

In [17]:
sp_df = df_scores[df_scores['Institute Name'].str.contains('Indian Institute of Technology')]
fig = sp_df.set_index('Institute Name').T.plot()
fig.update_layout(width=1000, height=500)
fig.show()

***Performance of NITs***

In [18]:
sp_df = df_scores[df_scores['Institute Name'].str.contains('National Institute of Technology')]
fig = sp_df.set_index('Institute Name').T.plot()
fig.update_layout(width=1000, height=500)
fig.show()

# **Machine Learning Model**

In [19]:
!pip install -U scikit-learn

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [20]:
# TF-IDF vectorization of text
from nltk.corpus import stopwords
from sklearn.feature_extraction import text 
from sklearn.feature_extraction.text import TfidfVectorizer
#from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS as sklearn_stop_words
#from sklearn.feature_extraction import _stop_words 
from sklearn.feature_extraction._stop_words import ENGLISH_STOP_WORDS

corpus = ml_dataset['Institute ID'].values

my_stop_words = ml_dataset['Institute ID'] 
stop_words = text.ENGLISH_STOP_WORDS.union(my_stop_words)
 
vectorizer = TfidfVectorizer(stop_words=None)
#vectorizer = TfidfVectorizer(stop_words=my_stop_words)
 
X = vectorizer.fit_transform(corpus)
 
print(vectorizer.get_feature_names_out())
 
# Visualizing the Document Term Matrix using TF-IDF
import pandas as pd
VectorizedText=pd.DataFrame(X.toarray(), columns=vectorizer.get_feature_names_out())
VectorizedText['originalText']=pd.Series(corpus)
VectorizedText

['0011' '0013' '0014' '0015' '0017' '0020' '0025' '0043' '0053' '0055'
 '0056' '0064' '0072' '0078' '0080' '0092' '0098' '0099' '0105' '0108'
 '0130' '0139' '0143' '0146' '0147' '0149' '0155' '0163' '0168' '0169'
 '0172' '0175' '0184' '0186' '0189' '0190' '0197' '0201' '0202' '0205'
 '0207' '0217' '0221' '0223' '0237' '0249' '0255' '0263' '0267' '0273'
 '0284' '0286' '0297' '0306' '0308' '0334' '0355' '0356' '0357' '0363'
 '0367' '0373' '0374' '0378' '0379' '0384' '0388' '0389' '0391' '0395'
 '0410' '0436' '0439' '0445' '0454' '0455' '0456' '0458' '0460' '0467'
 '0473' '0474' '0476' '0489' '0490' '0491' '0493' '0496' '0497' '0507'
 '0513' '0516' '0522' '0530' '0535' '0541' '0555' '0560' '0564' '0573'
 '0575' '0577' '0584' '0604' '0613' '0614' '0615' '0616' '0619' '0620'
 '0621' '0622' '0643' '0685' '0701' '0733' '0747' '0749' '0765' '0774'
 '0795' '0844' '0854' '0878' '0938' '10' '1015' '10342' '1074' '1075'
 '1093' '1119' '1262' '1269' '1297' '13' '1331' '1336' '1345' '1352'
 '1356' '

0011  0013  0014  0015  0017  0020  0025  0043  0053  0055  ...  45375  \
0     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
1     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
2     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
3     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
4     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
..    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...    ...   
195   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
196   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
197   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
198   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
199   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   

     46111  46239  50605  6379  6581  7252  8898        ir  originalText  
0      0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482   IR-E-U-0456  
1      0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482   IR-E-I-1074  
2      0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482   IR-E-U-0306  
3      0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482   IR-E-I-1075  
4      0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482   IR-E-U-0573  
..     ...    ...    ...   ...   ...   ...   ...       ...           ...  
195    0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482   IR-E-U-0854  
196    0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482  IR-E-C-33777  
197    0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482  IR-E-C-16504  
198    0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482   IR-E-C-1336  
199    0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482  IR-E-C-19582  

[200 rows x 202 columns]

In [21]:
VectorizedText.head()

0011  0013  0014  0015  0017  0020  0025  0043  0053  0055  ...  45375  \
0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
1   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
2   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
3   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
4   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   

   46111  46239  50605  6379  6581  7252  8898        ir  originalText  
0    0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482   IR-E-U-0456  
1    0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482   IR-E-I-1074  
2    0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482   IR-E-U-0306  
3    0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482   IR-E-I-1075  
4    0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482   IR-E-U-0573  

[5 rows x 202 columns]

In [22]:
DataForML = ml_dataset.join(VectorizedText)

In [23]:
DataForML.head()

Institute ID  Score  Rank  0011  0013  0014  0015  0017  0020  0025  ...  \
0  IR-E-U-0456  90.04     1   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
1  IR-E-I-1074  88.12     2   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
2  IR-E-U-0306  83.96     3   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
3  IR-E-I-1075  82.56     4   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
4  IR-E-U-0573  78.89     5   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   

   45375  46111  46239  50605  6379  6581  7252  8898        ir  originalText  
0    0.0    0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482   IR-E-U-0456  
1    0.0    0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482   IR-E-I-1074  
2    0.0    0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482   IR-E-U-0306  
3    0.0    0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482   IR-E-I-1075  
4    0.0    0.0    0.0    0.0   0.0   0.0   0.0   0.0  0.175482   IR-E-U-0573  

[5 rows x 205 columns]

In [24]:
DataForML.drop(['Institute ID', 'originalText'], axis=1, inplace=True)

*Splitting the dataset*

In [25]:
X = DataForML.drop(columns='Rank', axis = 1)
Y = DataForML['Rank']
print(X)
print(Y)

     Score  0011  0013  0014  0015  0017  0020  0025  0043  0053  ...  43708  \
0    90.04   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
1    88.12   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
2    83.96   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
3    82.56   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
4    78.89   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
..     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...    ...   
195  33.90   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
196  33.88   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
197  33.77   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
198  33.76   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   
199  33.70   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    0.0   

     45375  46111  46239  50605  6379  

*Data splitting for training and testing*

In [26]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.33, random_state=1)

*Model Creation*

In [27]:
model = LogisticRegression()

In [28]:
# training the LogisticRegression model with Training data
model.fit(X_train, Y_train)

/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning:

lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression



LogisticRegression()

# **Model Accuracy**

In [29]:
# accuracy on training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)
print('Accuracy on Training data : ', training_data_accuracy)

Accuracy on Training data :  0.8731343283582089


In [30]:
# accuracy on test data
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)
print('Accuracy on Test data : ', test_data_accuracy)

Accuracy on Test data :  0.0


*The model accuracy is around 87% which is appreciated for the amount of data. However, if higher accuracy is required, test_size and random_state can be manipulated to aim for higher accuracy.*